docs/writing-rules/data-flow/taint-mode.md

```diff
@@ -39,63 +43,208 @@ You can find more examples of taint rules in the [Semrep Registry](https://semgr
 
-```javascript
+<iframe src="https://semgrep.dev/embed/editor?snippet=eRvy" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+There are also [sanitizers by side effect](#sanitizers-by-side-effect). For a good example of use of both sources and sanitizers by side-effect, take a look at rule [`c.lang.security.use-after-free.use-after-free`](https://semgrep.dev/playground/r/c.lang.security.use-after-free.use-after-free).
+
+Propagators
+-----------
+
+By default, tainted data automatically propagates through assignments, operators, and function calls (from inputs to output). However, there are other ways taint can propagate, which may require language- or library-specific knowledge that Semgrep does not have built in.
+
+Consider the following Python code where an unsafe `user_input` is stored into a `set` data structure. A random element from `set` is then passed into a `sink` function. This random element can be `user_input` itself, leading to an injection vulnerability!
+
+```python
+def test(s):
+    x = user_input
+    s = set([])
+    s.add(x)
+    #ruleid: test
+    sink(s.pop())
+```
+
+The following rule cannot find the above-described issue. The reason is that Semgrep does not know that executing `s.add(x)` makes `x` one of the elements in the set data structure `s`.
+
+```yaml
+mode: taint
 pattern-sources:
-  - patterns:
-    - pattern-inside: function ($REQ, ...) {...}
-    - pattern: $REQ
+- pattern: user_input
+pattern-sinks:
+- pattern: sink(...)
+```
+
+The use of **taint propagators** enables Semgrep to propagate taint in this and other scenarios.
+Taint propagators are specified under the `pattern-propagators` key:
+
+```yaml
+pattern-propagators:
+- pattern: $S.add($E)
+  from: $E
+  to: $S
+```
+
+In the example above, Semgrep finds the pattern `$S.add($E)`, and it checks whether the code matched by `$E` is tainted. If it is tainted, Semgrep propagates that same taint to the code matched by `$S`. Thus, adding tainted data to a set marks the set itself as tainted.
+
+<iframe src="https://semgrep.dev/embed/editor?snippet=dGRE" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+Another situation where taint propagatos may be useful is to specify in Java that, when we iterate a collection that is tainted, the individual elements must also be considered tainted:
+
+```yaml
+pattern-propagators:
+- pattern: $C.forEach(($X) -> ...)
+  from: $C
+  to: $X
+```
+
+In general, a taint propagator must specify:
+1. A pattern containing **two** metavariables. These two metavariables specify where taint is propagated **from** and **to**.
+2. The `to` and `from` metavariables. These metavariables should match a **variable** or an **expression**.
+    - The `from` metavariable specifies the entry point of the taint.
+    - The `to` metavariable specifies where the tainted data is propagated to, typically an object or data structure.
```

Comment:
@IagoAbal If this is the most important information for the reader - which it seems so - move it under the `Propagators-----------` header.